src/main/java/org/javarosa/xform/parse/XFormParser.java

```diff
@@ -2898,14 +2895,9 @@ private void checkDependencyCycles() {
         }
 
         if (!acyclic) {
-            String errorMessage = "Logic Dependency Cycle:\n";
-            for (int i = 0; i < edges.size(); i++) {
-                TreeReference[] edge = (TreeReference[])edges.elementAt(i);
-                errorMessage += edge[0].toString() + " references " + edge[1].toString() + "\n";
-            }
-            reporter.error(errorMessage);
-
-            throw new RuntimeException("Logic is cyclical, referencing itself.");
+            ArrayList<String> cycleString = new ShortestCycleAlgorithm(edges).getCycle();
+            reporter.error("Logic is cyclical, referencing itself. Shortest Cycle " + cycleString);

```

Comment:
Relying on the toString of ArrayList here doesn't seem safe to me. I think we need a deliberate representation of what this should look like

Reply:
also: this message is duplicated and should be put in an intermediate String and re-used.